test/functional/wallet_reorgsrestore.py

```diff
@@ -90,6 +91,54 @@ def test_coinbase_automatic_abandon_during_startup(self):
         # Verify the coinbase descendant was also marked as abandoned
         assert_equal(wallet0.gettransaction(descendant_tx_id)['details'][0]['abandoned'], True)
 
+    def test_reorg_handling_during_unclean_shutdown(self):
+        self.log.info("Test that wallet doesn't crash due to a duplicate block disconnection event after an unclean shutdown")
+        node = self.nodes[0]
+        # Receive coinbase reward on a new wallet
+        node.createwallet(wallet_name="reorg_crash", load_on_startup=True)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        self.generatetoaddress(node, 1, wallet.getnewaddress(), sync_fun=self.no_op)
+
+        # Restart to ensure node and wallet are flushed
+        self.restart_node(0)
+        wallet = node.get_wallet_rpc("reorg_crash")
+        assert_greater_than(wallet.getwalletinfo()['immature_balance'], 0)
+
+        # Disconnect tip and sync wallet state
+        tip = wallet.getbestblockhash()
+        wallet.invalidateblock(tip)
+        wallet.syncwithvalidationinterfacequeue()
+
+        # Tip was disconnected, ensure coinbase has been abandoned
+        assert_equal(wallet.getwalletinfo()['immature_balance'], 0)
+        coinbase_tx_id = wallet.getblock(tip, verbose=1)["tx"][0]
+        assert_equal(wallet.gettransaction(coinbase_tx_id)['details'][0]['abandoned'], True)
+
+        # Abort process abruptly to mimic an unclean shutdown (no chain state flush to disk)
+        node.process.kill()
+
+        # Restart the node and confirm that it has not persisted the last chain state changes to disk
+        self.start_node(0)
+        assert_equal(node.getbestblockhash(), tip)
+
+        # Due to an existing bug, the wallet incorrectly keeps the transaction in an abandoned state, even though that's
+        # no longer the case (after the unclean shutdown, the node's chain returned to the pre-invalidation tip).
+        # This issue blocks any future spending and results in an incorrect balance display.
```

Comment:
>  This issue blocks any future spending and results in an incorrect balance display.

Technically correct but since the block was supposed to be invalidated and with the assumption that the likelihood for it to be invalidated again (successfully this time) is high, this immature unspent will not be blocked for long.

Not suggesting any changes, just noteworthy. Please do LMK if my assumption is incorrect.

Reply:
> Technically correct but since the block was supposed to be invalidated and with the assumption that the likelihood for it to be invalidated again (successfully this time) is high, this immature unspent will not be blocked for long.

Usually, yes.
Yet, there is another edge case. "Invalidated" means a reorg, which could also be reversed. That is, a block reorged out of the active chain before the unclean shutdown could become part of the active chain again when the node restarts (due to an extension of that chain - new block). In this scenario, the node starts at the reorged block, so no block connection event is signaled, and continues extending such chain. Leaving the tx in an invalid state forever (users can fix this state by calling `rescanblockchain`).